### ISSUE TITLE
Sitelet boxing/unboxing

### ISSUE THREAD

**Issue by Jand42** [dotnet-websharper/websharper#861]
Currently, `SiteletBuilder`, intended for C# use returns a `Sitelet<obj>` which type tests correctly for the types that are added by `.With<T>(...)`. These same type tests are missing when using `Sitelet.Upcast`, resulting in a `Sitelet<obj>` that is doing an unsafe downcast on action values for mapping them to an URL, this is wrong when the Sitelet is later combined with a `Sum`.

Todo: rename `Sitelet.Upcast` to `Sitelet.Box` and `Sitelet.UnsafeDowncast` to `Sitelet.UnboxUnsafe`, and make both operations safe.

**Comment by Jand42** [comment]
Tested: if you have two or more [<WebSite>] attributes (defined either on a static value or assembly level) in two different projects, Sitelet runtime tries to `Sum` them, which fails because of this unsafe boxing, with an error like `Unable to cast object of type 'Site' to type 'Action'.` where `Site` and `Action` are two different endpoint types.

With correct box/unbox, all sites would run in parallel, but the order in which some routers can shadow another depends on the order of assemblies returned by `BuildManager.GetReferencedAssemblies`. This behavior is not robust enough, creating new ticket about fixing this.
